ouroboros-consensus/test-storage/Test/Ouroboros/StateMachine.hs

```diff
@@ -0,0 +1,510 @@
+        in standardPostCondition (\ls1 ls2 -> fromList ls1 .== fromList ls2) r1 r2
+    (DoesDirectoryExist _ path , DoesDirectoryExistR r1) ->
+        let (r2, _) = runState (runExceptT $ mockDoesDirectoryExist2 path) fs
+        in standardPostCondition (.==) r1 r2
+    (DoesFileExist _ path , DoesFileExistR r1) ->
+        let (r2, _) = runState (runExceptT $ mockDoesFileExist2 path) fs
+        in standardPostCondition (.==) r1 r2
+    (Dummy, DummyR) -> Bot
+    _ -> Bot
+
+
+standardPostCondition :: (a -> b -> Logic) -> Either FsError a -> Either FsError b -> Logic
+standardPostCondition _ (Left e1) (Left e2) =
+    if e1 `sameFsError` e2 then Top
+    else Bot .// ("IO returned error " ++ show e1 ++ " while Sim returned " ++ show e2)
+standardPostCondition _ (Left e1) (Right b) = Bot .// "IO returned " ++ (show e1)
+standardPostCondition _ (Right a) (Left e2) = Bot .// "Sim returned " ++ (show e2)
+standardPostCondition checkEq (Right a) (Right b) = checkEq a b
+
+allTop :: a -> b -> Logic
+allTop _ _ = Top
+
+generator :: Model Symbolic -> Gen (Command Symbolic)
+generator m@(Model _ n _ mp) =
+    let openfds = Map.size mp
+        -- frequency of 'opens' reduces when openfds rises
+        -- and becomes 0 when there are 200 files to avoid EMFILE.
+        fOpen = div (max (200 - openfds) 0) 40
+        -- we want 0 probability when there are no open files.
+        fClose = if openfds == 0 then 0 else 1 + (div (max (openfds) 0) 25)
+        fFileExist = if openfds == 0 then 0 else 1
+        genFD :: Gen (Fd, Reference (Opaque (FsHandle IOFSE)) Symbolic)
+        genFD = do
+            (fd :: Fd) <- elements $ Map.keys mp
+            let x = snd (mp ! fd)
+            pure (fd, x)
+    in
+        frequency
+        [(fOpen, Open <$> return (n == 0) <*> genFile <*> genMode)
+        ,(fClose, Close <$> return (n == 0) <*> genFD)
+--        ,(fFileExist, Seek <$> return (n == 0) <*> genFD <*> genSeekMode <*> arbitrary)
+        ,(fFileExist, Get <$> return (n == 0) <*> genFD <*> suchThat arbitrary (>=0))
+        ,(fFileExist, Put <$> return (n == 0) <*> genFD <*> (BL.pack <$> arbitrary))
+        ,(1, CreateDirectory <$> return (n == 0) <*> genPath)
+        ,(1, ListDirectory <$> return (n == 0) <*> genPath)
+        ,(1, DoesDirectoryExist <$> return (n == 0) <*> genPath)
+        ,(1, DoesFileExist <$> return (n == 0) <*> genPath)
+        ,(1, return Dummy)
+        ]
+
+genPath :: Gen FsPath
+genPath =
+    let x = ["xx", "yy", "zz", "ww"]
+    in suchThat (sublistOf x) (\ls -> Prelude.length ls > 0)
+
+genFile :: Gen FsPath
+genFile =
+    let x = [["aa"], ["bb"], ["cc"], ["dd"]]
+    in elements x
+
+genMode :: Gen IOMode
+genMode = elements [ReadMode, AppendMode, ReadWriteMode]
+
+genSeekMode :: Gen SeekMode
+genSeekMode = elements [AbsoluteSeek, RelativeSeek, SeekFromEnd]
+
+shrinker :: Command r -> [Command r]
+shrinker _ = []
+
+shouldClean :: Command r -> Bool
+shouldClean cmd = case cmd of
```

Comment:
Same as above -- let's try to avoid this machinery, and instead take advantage of q-s-m init step.